### PR TITLE
Hankel

### PR DESCRIPTION
**Pull request by martin-ueding**
Continuation of https://github.com/HISKP-LQCD/hadron/pull/215 with additional fixes to make it pass all the checks.

### CODE REVIEW

**Comment by martin-ueding** [comment]
One needs to load the hadron library since in the testing environment it is not loaded in the Rprofile file.

**Comment by martin-ueding** [comment]
The vignette still does not run through:

Bildschirmfoto_002

I have tried to change the `id` parameter such that it would not trigger that condition, but that does not work either:

Bildschirmfoto_003

And also not into the other direction.


Bildschirmfoto_004

Guess there is something still off.

**Comment by urbach** [comment]
sorry, my bad, I hope I fixed it now!?

**Comment by martin-ueding** [comment]
Adding `library(hadron)` was straightforward. I don't understand the error in the vignette and therefore could not fix it yet.

**Comment by urbach** [comment]
But I fixed it, didn't I?

On 20 December 2019 16:24:01 CET, Martin Ueding <notifications@github.com> wrote:
>Adding `library(hadron)` was straightforward. I don't understand the
>error in the vignette and therefore could not fix it yet.

-- 
Carsten Urbach mobile

File: R/hankel.R [hunk 1]
```diff
@@ -102,65 +111,138 @@ gevp.hankel <- function(cf, t0=1, deltat=1, n, N, eps=0.0001, range=c(0,1),
 #' @param cf object of type \link{cf}
 #' @param t0     initial time value of the GEVP, must be in between 0 and
 #'    \code{Time/2-2}. Default is 1.
-#' @param deltat Integer. Time shift to be used to build the Hankel matrix
-#' @param n Integer.
+#' @param n Integer. Size of the submatrix Hankel matrices to generate
 #' @param N Integer.
 #' @param eps Numeric. Cut-off
```
Comment (kostrzewa) [inline R/hankel.R]: cut-off of what?

File: R/hankel.R [hunk 2]
```diff
@@ -102,65 +111,138 @@ gevp.hankel <- function(cf, t0=1, deltat=1, n, N, eps=0.0001, range=c(0,1),
 #' @param cf object of type \link{cf}
 #' @param t0     initial time value of the GEVP, must be in between 0 and
 #'    \code{Time/2-2}. Default is 1.
-#' @param deltat Integer. Time shift to be used to build the Hankel matrix
-#' @param n Integer.
+#' @param n Integer. Size of the submatrix Hankel matrices to generate
 #' @param N Integer.
```
Comment (kostrzewa) [inline R/hankel.R]: Integer describing what?

File: R/hankel.R [hunk 3]
```diff
@@ -102,65 +111,140 @@ gevp.hankel <- function(cf, t0=1, deltat=1, n, N, eps=0.0001, range=c(0,1),
...
-  }
+bootstrap.hankel <- function(cf, t0, n=2, N = cf$Time/2+1, id=c(1), range=c(0,1), eps=0.001) {
+
+  stopifnot(inherits(cf, 'cf_meta'))
+  stopifnot(inherits(cf, 'cf_boot'))
 
-  if(cf$boot.samples == FALSE) {
-    stop("Need to call bootstrap.cf first. Aborting...\n")
+  boot.R <- cf$boot.R
+  ##if(t0+2*n+deltat >= N) {
```
Comment (kostrzewa) [inline R/hankel.R]: debug stuff left

**Comment by urbach** [comment]
I hope I have fixed everything now!?

**Comment by martin-ueding** [comment]
The vignette now builds for me. Let's land this feature.

File: R/hankel.R [hunk 4]
```diff
@@ -43,7 +35,15 @@ gevp.hankel.old <- function(cf, t0, deltat = 1, n, N, id=c(1),
...
 #' @param debug Boolean. Enable debug output.
-#'
+#' @param deltat Integer. Time shift to be used to build the Hankel matrix
+#' @param submatrix.size Integer. Submatrix size to be used in build
+#'   of Hankel matrices. Submatrix.size > 1 is experimental.
+#' @param element.order Integer vector. specifies how to fit the \code{n} linearly ordered single
+#'    correlators into the correlator
+#'    matrix for submatrix.size > 1. \code{element.order=c(1,2,3,4)} leads to a matrix
+#'    \code{matrix(cf[element.order], nrow=2)}.
+#'    Double indexing is allowed.
```
Comment (kostrzewa) [inline R/hankel.R]: @urbach: what does "double indexing" mean?

Comment (martin-ueding) [reply R/hankel.R]: I guess that means you could give a triangular matrix and then specify `element.order = c(1, 2, 2, 3)`. If that is the case I'd rather call describe with a bit more prose and perhaps this example.

**Comment by urbach** [comment]
> > @@ -43,7 +35,15 @@ gevp.hankel.old <- function(cf, t0, deltat = 1, n, N, id=c(1),
>  #'
>  #' @inheritParams bootstrap.hankel
>  #' @param debug Boolean. Enable debug output.
> -#'
> +#' @param deltat Integer. Time shift to be used to build the Hankel matrix
> +#' @param submatrix.size Integer. Submatrix size to be used in build
> +#'   of Hankel matrices. Submatrix.size > 1 is experimental.
> +#' @param element.order Integer vector. specifies how to fit the \code{n} linearly ordered single
> +#'    correlators into the correlator
> +#'    matrix for submatrix.size > 1. \code{element.order=c(1,2,3,4)} leads to a matrix
> +#'    \code{matrix(cf[element.order], nrow=2)}.
> +#'    Double indexing is allowed.
> 
> I guess that means you could give a triangular matrix and then
> specify `element.order = c(1, 2, 2, 3)`. If that is the case I'd
> rather call describe with a bit more prose and perhaps this example.

That's one example of what I ment... Maybe one should better say
"Double usage of matrix elements is allowed"?

**Pull request merged**